### DISCUSSION TITLE
Downloading an Artifact that Does not Exist Yet

### DISCUSSION THREAD

**Discussion by naftulikay** [community/community#27041] [Actions]
<p>I have what I feel is a fairly simple situation where I would like to cache my Python dependencies within my <code>virtualenv</code> in <code>venv/</code>. I might look into <code>actions/cache@v2</code> for the <code>~/.cache/pip</code> dependencies, but for now, all I care about is caching everything in <code>venv/</code>.</p>
<pre><code class="lang-yaml">---
name: deploy
on:
  push:
    branches:
      - master
  pull_request:
    branches:
      - master
  workflow_dispatch:

jobs:
  build:
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v2
      - uses: actions/setup-python@v2
        with:
          python-version: 3.9.5
      - run: pip install virtualenv
      - uses: actions/download-artifact@v2
        with:
          name: pip
      - run: virtualenv venv/
      - run: source venv/bin/activate
      - run: pip install -r requirements.txt
      - uses: actions/upload-artifact@v2
        with:
          name: pip
          path: venv/
      - run: ansible --version
      - run: ansible-galaxy install -f -p .ansible/galaxy-roles -r requirements.yml
      - run: ansible-playbook -u ansible -C playbooks/configure.yml
        env:
          ANSIBLE_FORCE_COLOR: "true" 
</code></pre>
<p>In a nutshell, I want to download the <code>venv</code> folder if it exists, run a <code>pip install -r requirements.txt</code> to install/update dependencies if necessary, store <code>venv</code> folder as an artifact, and then continue on to the rest of my work.</p>
<p>I’ll eventually be splitting this up into its own job, maybe I’ll call it <code>prepare</code> and it will run before everything else so it can be decoupled, but that’s not urgent right now.</p>
<p>When I run <code>actions/download-artifact@v2</code>, I get an error that the artifact does not exist yet:</p>
<pre><code class="lang-auto">Error: Unable to find any artifacts for the associated workflow
</code></pre>
<p>It seems rather arbitrary to have to run my workflow once with <code>actions/download-artifact@v2</code> commented out and then run it again with it uncommented to get the cache populated once, and if this build does not run for 90 days, it will presumably fail as the artifacts expire.</p>
<p>I am new to GitHub actions: what is the design pattern for doing this? I feel like I’m trying to do something fairly simple but running into a roadblock, and there’s no <code>if-no-files-found</code> directive on <code>actions/download-artifact@v2</code>.</p>
<p>How do I try to restore an artifact and ignore errors if it does not already exist?</p>

**Comment by jsoref** [comment, marked as answer]
<p>Technically you can cheat and use:</p>
<pre><code class="lang-auto">if: always()
</code></pre>
<p>to force your later steps to run. This way your first run might technically “fail”, but it will “prime” your fake cache for the second run.</p>
<p>Formally, github clearly wants you to use <code>actions/cache</code> instead.</p>

**Comment by naftulikay** [comment]
<aside class="quote no-group" data-username="jsoref" data-post="2" data-topic="187652">
<div class="title">
<div class="quote-controls"></div>
 jsoref:</div>
<blockquote>
<p>Formally, github clearly wants you to use <code>actions/cache</code> instead.</p>
</blockquote>
</aside>
<p>Yes, this is the idea and was the solution. I was confused as to how <code>actions/cache</code> worked, I didn’t understand how inserting it at one step would accomplish <em>storing</em> the cache at the end of the run, but apparently it hooks into the entire run and saves the cache at the end.</p>
<p>Solution:</p>
<pre><code class="lang-yaml">---
name: deploy
on:
  push:
    branches:
      - master
  pull_request:
    branches:
      - master
  workflow_dispatch:

jobs:
  build:
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v2

      - uses: actions/setup-python@v2
        with:
          python-version: 3.9.5

      - uses: actions/cache@v2
        env:
          cache-name: python
        with:
          key: "${{ runner.os }}-build-${{ env.cache-name }}-${{ hashFiles('requirements.txt') }}"
          restore-keys: |
            ${{ runner.os }}-build-${{ env.cache-name }}
            ${{ runner.os }}-build
            ${{ runner.os }}
          path: |
            ~/.cache/pip
            ./venv/

      - run: pip install virtualenv
      - run: virtualenv venv/
      - run: source venv/bin/activate
      - run: pip install -r requirements.txt
      - run: ansible --version
      - run: ansible-galaxy install -f -p .ansible/galaxy-roles -r requirements.yml
      - run: ansible-playbook -u ansible --private-key .private/ssh-keys/ansible.key playbooks/configure.yml
        env:
          ANSIBLE_FORCE_COLOR: "true"
</code></pre>

**Comment by jsoref** [comment]
<p>Formally you want the cache early:</p>
<aside class="onebox githubblob" data-onebox-src="https://github.com/actions/cache/blob/0638051e9af2c23d10bb70fa9beffcad6cff9ce3/action.yml#L22-L24">
  <header class="source">

      <a href="https://github.com/actions/cache/blob/0638051e9af2c23d10bb70fa9beffcad6cff9ce3/action.yml#L22-L24" target="_blank" rel="noopener nofollow ugc">github.com</a>
  </header>

  <article class="onebox-body">
    <h4><a href="https://github.com/actions/cache/blob/0638051e9af2c23d10bb70fa9beffcad6cff9ce3/action.yml#L22-L24" target="_blank" rel="noopener nofollow ugc">actions/cache/blob/0638051e9af2c23d10bb70fa9beffcad6cff9ce3/action.yml#L22-L24</a></h4>



  <pre class="onebox">    <code class="lang-yml">
      <ol class="start lines" start="22" style="counter-reset: li-counter 21 ;">
          <li>main: 'dist/restore/index.js'</li>
          <li>post: 'dist/save/index.js'</li>
          <li>post-if: 'success()'</li>
      </ol>
    </code>
  </pre>


  </article>

  <div class="onebox-metadata">
    
    
  </div>

  <div style="clear: both"></div>
</aside>

<p>Specifically, it relies on <code>post</code> to perform the cache writes.</p>
<aside class="onebox githubblob" data-onebox-src="https://github.com/actions/cache/blob/0638051e9af2c23d10bb70fa9beffcad6cff9ce3/examples.md#python---pip">
  <header class="source">

      <a href="https://github.com/actions/cache/blob/0638051e9af2c23d10bb70fa9beffcad6cff9ce3/examples.md#python---pip" target="_blank" rel="noopener nofollow ugc">github.com</a>
  </header>

  <article class="onebox-body">
    <h4><a href="https://github.com/actions/cache/blob/0638051e9af2c23d10bb70fa9beffcad6cff9ce3/examples.md#python---pip" target="_blank" rel="noopener nofollow ugc">actions/cache/blob/0638051e9af2c23d10bb70fa9beffcad6cff9ce3/examples.md#python---pip</a></h4>


    <pre><code class="lang-md"># Examples

- [C# - NuGet](#c---nuget)
- [D - DUB](#d---dub)
  - [POSIX](#posix)
  - [Windows](#windows)
- [Elixir - Mix](#elixir---mix)
- [Go - Modules](#go---modules)
  - [Linux](#linux)
  - [macOS](#macos)
  - [Windows](#windows-1)
- [Haskell - Cabal](#haskell---cabal)
- [Java - Gradle](#java---gradle)
- [Java - Maven](#java---maven)
- [Node - npm](#node---npm)
  - [macOS and Ubuntu](#macos-and-ubuntu)
  - [Windows](#windows-2)
  - [Using multiple systems and `npm config`](#using-multiple-systems-and-npm-config)
- [Node - Lerna](#node---lerna)
- [Node - Yarn](#node---yarn)
</code></pre>


  This file has been truncated. <a href="https://github.com/actions/cache/blob/0638051e9af2c23d10bb70fa9beffcad6cff9ce3/examples.md#python---pip" target="_blank" rel="noopener nofollow ugc">show original</a>

  </article>

  <div class="onebox-metadata">
    
    
  </div>

  <div style="clear: both"></div>
</aside>

**Comment by Alex-ley-scrub** [comment]
@naftulikay / @jsoref isn't this the simplest solution? (I realise that it may have not been available in 2021 at the time of the original post):
```yaml
  - name: download pip artifact
  - continue-on-error: true
  - uses: actions/download-artifact@v6
  - with:
      name: pip
```

**Reply by jsoref** [reply]
You have too many `-`s (you only want the first one).

These days, you'd need [`overwrite: true`](https://github.com/actions/upload-artifact/?tab=readme-ov-file#overwriting-an-artifact) too which wasn't a thing until [`v4`](https://github.com/actions/upload-artifact/commit/11ff42c7b1b52130283d8a02bc960d1e1de95000).

Beyond that, in order to use artifacts, you'd probably eventually want to look into [`artifact-ids`](https://github.com/actions/download-artifact/?tab=readme-ov-file#download-artifacts-by-id), and I'd personally use [`gh run download`](https://cli.github.com/manual/gh_run_download) to retrieve artifacts instead of the action as you can also use [`gh run list`](https://cli.github.com/manual/gh_run_list) and friends giving a lot more flexibility than [`actions/download-artifact`](https://github.com/actions/download-artifact).
